Fix String with escapes prefix matching on Erlang.

aznashwan opened the pull request:
Fixes gleam-lang/gleam#3126.

Fixes String prefix matching on Erlang when the prefix contains escape characters or unicode codepoints by ensuring the correct prefix byte length is matched against.

lpil reviewed (commented):
Hello! Thank you for this

Could you reuse the code here please? 🙏  https://github.com/gleam-lang/gleam/blob/2c59b9c3f9b00fe0c35726e00e97f7843eff5c1f/compiler-core/src/javascript/pattern.rs#L856

It'll need to be make slightly more flexible to specify whether it's utf8 (Erlang) or utf16 (JavaScript)

aznashwan commented:
> Could you reuse the code here please?

Done, hadn't noticed it before since I only investigated the Erlang part of the compiler tree but it was basically exactly what I needed.

I've adapted the function to directly return the updated string, properly treat the escape chars Gleam supports, and moved it to a [top-level `crate::strings` module](https://github.com/gleam-lang/gleam/pull/3139/files#diff-c7e6b4e895ded4ed39bf97d94cb55c4c7561637ae8152f76f1e1bbdd2b2228a3R6) for lack of a better place to put it, but please feel free to suggest a new one.

I've also split the specific prefixes tested against to make the transpiled code a little easier to follow.

lpil reviewed (approved):
Lovely! Thank you.
